tasks/src/main/scala/dagr/tasks/ScatterGather.scala

```diff
@@ -0,0 +1,175 @@
+  }
+
+
+  /**
+   * Represents a gather operation in a scatter/gather pipeline.
+   */
+  sealed trait Gather[A, B <: Task] extends Task {
+    private[ScatterGather] var sources: Option[Seq[A]] = None
+  }
+
+  /** Implementation of a Gather that performs a single All->One gather operation. */
+  private class SingleGather[A,B <: Task](f: Seq[A] => B) extends Gather[A,B] {
+    override def getTasks: Traversable[_ <: Task] = sources match {
+      case None    => throw new IllegalStateException("Gather.getTasks called before sources populated.")
+      case Some(a) => Some(f(a))
+    }
+  }
+
+
+  /**
+  * Implementation of a Scatter that is just a thinly veiled wrapper around the Scatterer
+  * being used to generate the set of scatters/partitions to operate on.
+  */
+  private class ScattererWrapper[A](scatterer: Scatterer[A]) extends Scatter[A] {
+    override def getTasks: Traversable[_ <: Task] = Some(scatterer)
+
+    override def gather[B <: Task](f: (Seq[A]) => B): Gather[A,B] =
+      throw new UnsupportedOperationException("gather not supported on an unmapped Scatter")
+
+    override def map[T <: Task](f: A => T): Scatter[T] = {
+      val sub = new PrimarySubScatter(scatterer, f)
+      this ==> sub
+      sub
+    }
+  }
+
+  /**
+  * Sub-trait of scatter that exists solely because we need two sub-types that share some common functionality.
+  * Provides the tracking of tasks generated by the Scatter, sub-Scatters that need to be invoked and wired in
+  * when the partitions are available and the set of Gathers to be performed on this stage of the Scatter.
+  */
+  private trait SubScatter[A, B <: Task] extends Scatter[B] {
+    val tasks   = new ListBuffer[B]
+    val subs    = new ListBuffer[SubScatter[B, _ <: Task]]
+    val gathers = new ListBuffer[Gather[B, _ <: Task]]
+
+    /** Defined here for convenience, so it can be invoked on all subs, though in reality only ever called on Secondaries. */
+    def chain(a: A): B = ???
+
+    /** Ensures that gathers are wired in correctly. */
+    protected def connectGathers: Unit = {
+      gathers.foreach(gather => {
+        gather.sources = Some(tasks)
+        tasks.foreach(task => task ==> gather)
+      })
+
+      subs.foreach(_.connectGathers)
+    }
+
+    override def map[T <: Task](f: (B) => T): Scatter[T] = {
+      val sub = new SecondarySubScatter(f)
+      this.subs += sub
+      sub
+    }
+
+    override def gather[G <: Task](f: (Seq[B]) => G): Gather[B,G] = {
+      val gather = new SingleGather[B,G](f)
+      this.gathers += gather
+      gather
+    }
+  }
+
+  /**
+  * A type of Scatter that is used to represent the first round of Scatter operation after the
+  * scatterer has created the partitions.  It's tasks are generated using the scatter partitions
+  * as input, and is also responsible for chaining together sub-scatters.
+  */
+  private class PrimarySubScatter[A, B <: Task](val scatterer: Scatterer[A], f: A => B) extends Pipeline with SubScatter[A,B] {
```

Comment:
I've made it a `private val`.  FWIW if you access a constructor parameter anywhere other than in the constructor, scala promotes it to a val anyway, so I prefer to declare it that way.